### ISSUE TITLE
Queue Building Block API

### ISSUE THREAD

**Issue by davidfowl** [dapr/dapr#771]
## In what area(s)?

/area runtime

## Describe the feature

This could probably be built on top of the existing primitives but it feels like there should be a queue primitive as part of the runtime. There's pub-sub and there should be a queue model as well. Unlike pub-sub, the queue wouldn't broadcast to all subscribers of a topic but only a single worker can process a message on a queue at once, in the order those jobs were posted (FIFO).

**Comment by RicardoNiepel** [comment]
As long, as all subscribers are using the same app Id, only one will receive each message.

Which scenario do you have in mind where different Dapr apps (different Id's) are competing consumers on one topic?

**Comment by abhirockzz** [comment]
From what I understand, both "competing consumer" (queue behavior) and "broadcast" (topic behavior) can be implemented via a Dapr pub/sub component i.e. it depends on the implementation. E.g. for a Kafka pub/sub, all Dapr app instances will define the same "consumer group", therefore the message processing will be balanced. Same goes for the NATS implementation (its a binding not pub/sub but similar concept) which makes uses a queue subscription along with a queue group name

**Comment by georgestevens99** [comment]
Perhaps there is a fundamental concept relevant to distributed systems that is missing in thinking about the nature of queues, Pub/Subs and components.  

Namely, there are at least 2 kinds of communication "channels" in Dist Sys -- Reliable and unreliable.  RPC is generally unreliable since it depends upon the receiver being alive and well to receive the message, and if the receiver does not receive the message it is lost.  Note this is tightly temporally coupled (coupled in time).  

Reliable communication depends on an intermediary receiving the message (like a queue or a reliable topic) and then holding it until the intended receiver grabs it, stores it (or processes it), and dequeues it (often in that order to ensure reliability).  Skipping over the details, please note that reliable communications like queues are temporarily decoupled -- The received does not have to be alive at the time the message is received to actually grab it.  Such temporal decoupling provides a massive amount of resilience to distributed systems and off loads the sender from having to worry about what to do if the receiver is not alive to receive the message -- simplifying crash situations.  Also note that due to this RPC unreliable communications are much more open to cascading failure scenarios, where one component crashes, causing other components to become sketchy, etc, down the chain of communications.  This is not Anti-Fragile behavior.  Rather just the opposite.  

So, by focusing more upon the essential nature of reliable communications vs unreliable, you will be better able to design and promote the use of both Queues and Queued Topics in Dapr, in addition to Pub/Subs (which can operate on both reliable and unreliable communication channels, hopefully with a caviet emptor (sp) to the individual making that choice).  Clemens Vasters recently published a good blog on queues, FYI, at https://vasters.com/blog/.  Thanks, George

**Comment by georgestevens99** [comment]
To continue my previous line of thought, this time clearly stating to you what I value most for my work -- I need a high degree of reliability and resilience in messaging (both data transmission and  remote command invocation).  Specifically I have used a "double queued pub/sub" to furnish this in the past, and would love it if Dapr can provide such a thing, or at least the capability to compose such a thing out of Dapr and 3rd party components.  To me, a double queued pub/sub capability is a way, way higher priority than a queue component in Dapr.  

Specifically,  the double queued pub/sub consists of a pub/sub component (which Dapr already has) that runs in its own process.  Then there is the input queue to the pub/sub component, which also runs in a separate process.  And finally, the pub/sub component publishes to one or more  output queues (or maybe topics) that run in processes separate from the pub/sub component, and even separate from the input queue component.  Thus, with each of these parts (input queue, pub/sub component, output queues) running in their own processes the overall composition of components is highly reliable and resilient, making the loss of messages rare, or perhaps even close to impossible if well implemented.  Perhaps this capability already exists with Dapr's components and bindings.  Pardon me for not yet getting up to speed here, but that time is getting closer.

In summary, the highest priority for me is very high reliability and resilience of components and compositions of components that can be used in mission critical systems "out of the box".  Such usage will likely be beyond a single cluster, including on the Edge and even in  across-the-planet communications scenarios.   Thanks, George.

**Comment by databasedav** [comment]
> From what I understand, both "competing consumer" (queue behavior) and "broadcast" (topic behavior) can be implemented via a Dapr pub/sub component i.e. it depends on the implementation. E.g. for a Kafka pub/sub, all Dapr app instances will define the same "consumer group", therefore the message processing will be balanced (for events sent to the same topic). Same goes for the NATS pubsub implementation which uses a queue subscription along with a queue group name

so if we use NATS, is there a way to conditionally get broadcast behavior even if the app-id's are the same? I need both queue behavior for some services and broadcast for others; can I not use NATS? @abhirockzz

**Comment by tuapuikia** [comment]
AWS SNSSQS support FIFO queue. I will test it and get back to you on the result.
